### PR TITLE
8223373: Remove IntelliJ IDEA specific files from the source code repository

### PR DESCRIPTION
**Pull request by tsayao**
This PR does:

- Remove specific Idea files and let it be imported from gradle;
- Adds checkstyle (to use with checkstyle plugin - it will let you know style mistakes);
- Configures auto-format to sun style (with the changes mentioned in [Code Style Rules](https://wiki.openjdk.org/display/OpenJFX/Code+Style+Rules));
- Automatically sets Copyright notice (updates year too);

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8223373](https://bugs.openjdk.org/browse/JDK-8223373): Remove IntelliJ IDEA specific files from the source code repository

### Reviewers
 * [Marius Hanl](https://openjdk.org/census#mhanl) (@Maran23 - Committer)
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1009/head:pull/1009` \
`$ git checkout pull/1009`

Update a local copy of the PR: \
`$ git checkout pull/1009` \
`$ git pull https://git.openjdk.org/jfx.git pull/1009/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1009`

View PR using the GUI difftool: \
`$ git pr show -t 1009`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1009.diff">https://git.openjdk.org/jfx/pull/1009.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1009#issuecomment-1407752304)

### CODE REVIEW

**Comment by tsayao** [comment]
I would appreciate feedback from intellij idea users.

**Comment by tsayao** [comment]
CheckStyle plugin is also available for [Eclipse](https://checkstyle.org/eclipse-cs/#!/) if anyone is willing to set it up.

It caught many simple mistakes on the code.

@andy-goryachev-oracle You might like it :)

**Comment by andy-goryachev-oracle** [comment]
will take a look at the CheckStyle plug-in, thanks!

**Comment by andy-goryachev-oracle** [comment]
Since you are touching /.gitignore, could you also

- Add hs_err*.log, .DS_Store to .gitignore

from JDK-8296810 ?

**Comment by tsayao** [comment]
I would like to add manual tests to gradle so it would be imported by the IDE in a way it doesn't build (most are single java files that can be run directly) but code completion works. Does anyone know how to do that?

**Comment by Maran23** [comment]
Since I also use IntelliJ, I will review the PR in the next few days. :)

**Comment by tsayao** [comment]
I still cannot run manual tests directly from the IDE. I think it's close to work, but something is missing and I'm not a gradle ninja.

**Comment by tsayao** [comment]
Eclipse outputs the same error when running manual tests, so I think it's not possible to fix without restructuring the folder. 

Ready for review.

**Comment by kevinrushforth** [comment]
This touches `build.gradle` and `settings.gradle` in addition to IDE-specific files. Those changes will need additional review, so I'm requesting two reviewers.

/reviewers 2

File: .idea/runConfigurations/Ensemble8_jar.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<component name="ProjectRunConfigurationManager">
```
Comment (Maran23) [inline .idea/runConfigurations/Ensemble8_jar.xml:1]: Do we really need these run configurations (except `Run_Robot_Tests.xml`, which is useful)

File: .idea/misc.xml [hunk 2]
```diff
@@ -1,8 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
   <component name="ExternalStorageConfigurationManager" enabled="true" />
+  <component name="FrameworkDetectionExcludesConfiguration">
+    <file type="web" url="file://$PROJECT_DIR$" />
+  </component>
   <component name="IdProvider" IDEtalkID="9BC0DE4049720C1246C2D3B835C6F7C3" />
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_11" project-jdk-name="11" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_17" project-jdk-name="19" project-jdk-type="JavaSDK">
```
Comment (Maran23) [inline .idea/misc.xml:8]: I think this should be `project-jdk-name="17"`. When I check out this PR, IntelliJ also changed that automatically.

File: .idea/runConfigurations/Run_Robot_Tests.xml [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<component name="ProjectRunConfigurationManager">
+  <configuration default="false" name="Run Robot Tests" type="GradleRunConfiguration" factoryName="Gradle">
+    <ExternalSystemSettings>
+      <option name="executionName" />
+      <option name="externalProjectPath" value="$PROJECT_DIR$" />
+      <option name="externalSystemIdString" value="GRADLE" />
+      <option name="scriptParameters" value="-PFULL_TEST=true -PUSE_ROBOT=true" />
```
Comment (Maran23) [inline .idea/runConfigurations/Run_Robot_Tests.xml:7]: Is this also possible for manual tests? Or how do you launch manual tests?

File: .idea/codeStyles/Project.xml [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<component name="ProjectCodeStyleConfiguration">
+  <code_scheme name="Project" version="173">
```
Comment (Maran23) [inline .idea/codeStyles/Project.xml:2]: We should consider adding `<option name="LINE_SEPARATOR" value="&#10;" />` -> \n as separator

Comment (Maran23) [inline .idea/codeStyles/Project.xml:2]: Maybe `OpenJFX` is a better name here?

**Review by Maran23 — COMMENTED** [review]
Looks good to me. I just tested it with the newest `IntelliJ` version (2022.3.2) and it works fine.
It also follows what IntelliJ is suggesting. 
Also the gradle changes looks fine to me, although not sure how I can start the manual tests.
I left some comments.

File: .idea/codeStyles/Project.xml (hunk 4 above)
Comment (tsayao) [reply .idea/codeStyles/Project.xml:2]: I tried to change it but didn't work. It seems IntelliJ expects "Project" there.

Comment (tsayao) [reply .idea/codeStyles/Project.xml:2]: Added.

File: .idea/runConfigurations/Run_Robot_Tests.xml (hunk 3 above)
Comment (tsayao) [reply .idea/runConfigurations/Run_Robot_Tests.xml:7]: I do it by clicking on the run icon of the main method. Manual tests uses non-standard folder structure, so I had to customize the gradle script. As they are not part of the build, it's not possible (I think) to create run configurations.

File: .idea/runConfigurations/Ensemble8_jar.xml (hunk 1 above)
Comment (tsayao) [reply .idea/runConfigurations/Ensemble8_jar.xml]: Yeah, I agree. Removed those and left the robot tests.

File: .idea/misc.xml (hunk 2 above)
Comment (tsayao) [reply .idea/misc.xml:8]: Each user probably would have to configure a JDK for the project.

Comment (Maran23) [reply .idea/misc.xml:8]: Yes I agree. But since the language level is JDK-17 and this file is part of the version control, I would recommend we synchronize the `project-jdk-name` with the `languageLevel`, in this case `17`.

Comment (tsayao) [reply .idea/misc.xml]: I've changed it to "17".

**Review by Maran23 — APPROVED** [review]
Looks good to me. Tested with the latest IntelliJ version, everything works as expected.
The addition to the `.gitignore` is also very helpful and I can confirm I get no unrelated IntelliJ/Sonarlint files anymore.
Gradle changes also looks fine as far as I can tell. But someone else should probably have a look as well.

**Comment by tsayao** [comment]
Intellij Ideas does not see the dependencies:
```
dependencies {
        testImplementation project(":graphics").sourceSets.test.output
        testImplementation project(":base").sourceSets.test.output
        testImplementation project(":controls").sourceSets.test.output
        testImplementation project(":swing").sourceSets.test.output
    }
```

How do I rewrite it?

**Comment by Maran23** [comment]
> Intellij Ideas does not see the dependencies:
> 
> ```
> dependencies {
>         testImplementation project(":graphics").sourceSets.test.output
>         testImplementation project(":base").sourceSets.test.output
>         testImplementation project(":controls").sourceSets.test.output
>         testImplementation project(":swing").sourceSets.test.output
> }
> ```
> 
> How do I rewrite it?
> 
> It works if i remove `.sourceSets.test.output`, so I thought maybe It can be rewritten so it's good for Intellij and the build.

I do not know either. But with this PR the setup works already better than before and finally no more files inside my git staging.
So if you agree we can also just merge the IntelliJ changes now and try to fix the remaining issue(s) later, since this PR already improves the situation a lot. :)

**Comment by tsayao** [comment]
I agree. I have removed the gradle changed so it's easier to review.

**Review by Maran23 — APPROVED** [review]
LGTM, tested successfully on my Windows machine with IntelliJ.

**Comment by tsayao** [comment]
@kevinrushforth I've reverted the gradle changes, so it pretty much only affect the IDE.

File: .gitignore [hunk 5]
```diff
@@ -16,13 +16,55 @@ testbin/
...
+.idea/*.iml
+.idea/modules
+*.iml
+*.ipr
+
+# File-based project format
+*.iws
+
+# IntelliJ
+out/
```
Comment (kevinrushforth) [inline .gitignore:57]: This would pick up more than just IntelliJ files. Are they all needed?

File: .gitignore [hunk 6]
```diff
@@ -16,13 +16,55 @@ testbin/
...
+
+# SonarLint plugin
+.idea/sonarlint/
+#IntelliJ Idea End
+
+#Crash log
+hs_err*.log,
+
+#mac folder
+.DS_Store
```
Comment (kevinrushforth) [inline .gitignore:67]: These seem unrelated to this PR (and in the case of `hs_err`, there is a typo).

File: checkstyle.xml [hunk 7]
```diff
@@ -0,0 +1,213 @@
+<?xml version="1.0"?>
```
Comment (kevinrushforth) [inline checkstyle.xml:1]: I'm not keen on adding yet another top-level file in our repo. Can it go somewhere else?

**Review by kevinrushforth — COMMENTED** [review]
> I've reverted the gradle changes, so it pretty much only affect the IDE.

Good. I'll let the IntelliJ users do the review. My only comments are on the two files outside the `.idea` directory.

File: .gitignore (hunk 5 above)
Comment (Maran23) [reply .gitignore:57]: `out` is the default `Compiler output` path in IntelliJ. But I don't think it is really needed.

Comment (tsayao) [reply .gitignore]: Updated it, `*.iml` and `*.ipr` are the Idea project files that are created when importing from gradle.

File: .gitignore (hunk 6 above)
Comment (tsayao) [reply .gitignore]: Removed them

File: checkstyle.xml (hunk 7 above)
Comment (tsayao) [reply checkstyle.xml:1]: I've looked for a place, but didn't found one that would make sense. It could be used on eclipse as well and probably on netbeans.

Comment (tsayao) [reply checkstyle.xml:1]: @kevinrushforth can you guide me what should I do about this file?

Comment (kevinrushforth) [reply checkstyle.xml:1]: Since we already have a few other build and IDE-related files in the top repo, and since it could be shared by other IDEs, I won't object to putting it in the top repo.

**Comment by tsayao** [comment]
@arapte Would you take a look when you have time? Thanks.

**Comment by tsayao** [comment]
> I am testing with IntelliJ IDEA 2023.1.2 (Community Edition). Made some changes in a java file, added empty line with spaces, trailing spaces, added tabs at random places. All the spaces on empty line, trailing spaces were correctly removed, tabs were replaced with spaces by the IDE with or **without** this change. Am I missing anything here? Copyright year was not updated in the modified file. Are there any other steps to follow ? Also, without this change, the gradle tries to build the project as soon as we open the project in IDE, but with this change it does not. which seems OK to me.
> 
> Also I do not use IDE to build jfx nor to execute the tests. If you do, can you please share me any specific steps.

I configured it to follow Sun style with JavaFX exceptions (like allow longer line with). Converting tabs to space and removing trailing spaces are probably default.

About updating copyright year, maybe it's an Ultimate feature that might not be available in Community.

Tests are not working yet, because Idea is not able to correctly import from Gradle without changing it (I tried it but reverted to keep it unchanged for this PR).

**Review by arapte — APPROVED** [review]
Looks good.
Did not face any problem when loading the project in idea, changing properties, editing files.
